### ISSUE TITLE
Allow admins to configure flood control (failed login attempts)

### ISSUE THREAD

**Issue by herbdool** [backdrop/backdrop-issues#3110]
I was just reading about brute force login protection and then realized it's not all that clear what Backdrop does. This is inherited from Drupal where there were hidden settings and code comments explaining how it worked. It would be useful to allow admins to configure them.

In Drupal 7 there was [Flood Control](https://www.drupal.org/project/flood_control) which is a just a form to allow admins to change already existing settings. (The module actually also allows admin to configure the contact form thresholds).

And as a follow-up issue we could also incorporate [Flood Unblock](https://www.drupal.org/project/flood_unblock) to allow admins to unblock someone easily without having to go into the database.

---

PR by @herbdool: https://github.com/backdrop/backdrop/pull/2187

**Comment by quicksketch** [comment]
The PR at https://github.com/backdrop/backdrop/pull/2187 looks pretty good. I think I would stay away from exposing the text "Flood control" in the UI, that's a rather technical term for rate limiting. 

This looks like it would fold in nicely with https://github.com/backdrop/backdrop-issues/issues/2990, which suggests we add a settings tab to the contact module.

As for the new user flood settings, I think it would make sense if we made a new form for "Login settings" that was separate from "Account settings".

Overall, I'm not sure where this lands in common functionality. I've never encountered a problem with users getting locked out entirely. Usually I try a password reset if I get locked out (on any site, not just Backdrop sites) and that effectively works to circumvent the login lockout. The contact form rate limiting seems entirely reasonable though.

**Comment by quicksketch** [comment]
@herbdool Could you separate out the Contact module portion and combine it with https://github.com/backdrop/backdrop-issues/issues/2990, as a new PR for that issue? Then we can leave this issue for User Login flood control by itself.

**Comment by herbdool** [comment]
@quicksketch I've update the PR: took out contact module part (@opi already added it to his PR). And I changed the title and path. It's now at `admin/config/people/login` where we can add other settings as well.

I think one of the benefits of this is that it makes visible what was mysterious before: I don't think I understood how Drupal was blocking users.

**Comment by herbdool** [comment]
Updated with @quicksketch idea of making them inline from https://github.com/backdrop/backdrop-issues/issues/2990. Instead of using a theme function, I added a "container-inline" class to a fieldset to wrap the elements. I tried to limit the number of fieldsets since by default the Seven theme always puts borders around them. This might work for the contact form too @opi.

![screenshot from 2018-05-16 13-04-09](https://user-images.githubusercontent.com/1071483/40132037-aadeb542-5909-11e8-8201-8335da70237c.png)

**Comment by herbdool** [comment]
I figured out a better approach. I forgot about the container element. So I've replaced the fieldset with container so it now looks like:

![screenshot from 2018-05-16 13-36-17](https://user-images.githubusercontent.com/1071483/40133550-21ebe7be-590e-11e8-94a9-5f3b66a523d4.png)

Pretty much the same but now we don't get the help clearing by accident. And can use description again because the element is outside the inline container.

**Comment by quicksketch** [comment]
This looks pretty good. A few more suggestions:

- Can we move `user_login_method` out of `user_admin_settings()` so log in settings are all in the same place? Either that or this page should be labeled "Advanced login settings".
- "Login vs. log in"? Like "setup vs. set up", a "login" is a noun, and "log in" is a verb. Given the way we're using this, I think two words is correct in all cases here.
- Instead of a checkbox for "Use only the User ID to identify users", we could use radio buttons so it's clear what the unchecked version is. I suggest something like this:

```
Attempt identifier
(*) User ID and IP address combination (less secure, less likely to lock out users)
( ) User ID only (more secure, more likely to lock out users)
```

**Comment by herbdool** [comment]
@quicksketch I've updated the PR.

**Comment by opi** [comment]
Reviewed and tested

**Comment by klonos** [comment]
> I think one of the benefits of this is that it makes visible what was mysterious before: I don't think I understood how Drupal was blocking users.

👍 ...I had a support request this week and I told the customer that "I don't think that Drupal does this out of the box" ...that coming from a (supposedly) seasoned Drupal user 😞

**Comment by klonos** [comment]
Just reviewed the new page. Here are some suggestions to improve labels and help text:

1. We have "Users may log in using", but then "How to identify users attempting to log in". We should be more consistent. Lets either make these both questions or just statements. So either this:

    - Users may log in using:
    - Identify users attempting to log in by:

    ...or this:

    - How to allow users to log in
    - How to identify users attempting to log in

    I personally prefer the first. We should also make this text stand out more as a label. Perhaps bold?

2. Change "LIMIT LOGIN ATTEMPTS ..." to "LIMIT *FAILED* LOGIN ATTEMPTS ..."

3. We have:

    - ...BY IP ADDRESS
    - ...BY A USER

   ...either use A/AN for both, or not.

4. "...login attempts from one IP address per..." should perhaps be "...*failed* login attempts from *each* IP address per..."

5. Per-IP help/description:

    > Do not allow further login attempts from the current user's IP, if the limit has been reached. This is independent of the per-user limit below, and it helps to catch malicious login attempts from one IP using different user accounts. By default we have a reasonably high limit, since there may be only a single apparent IP for all users within the same institution.

    This text draws to much attention. Perhaps style it as help text (gray color).

6. "...login attempts by a user per..." -> "...*failed* login attempts by a user per..."

7. Per-user help/description:

    > Do not allow further login attempts from the current user, if the limit has been reached.

    Also style this to have gray color.

8. "(less secure, less likely to lock out users)" and "(more secure, more likely to lock out users)" ...can these be help text for each option? ...also is it "...to lock out users" or "...to lock users out"?

**Comment by herbdool** [comment]
@klonos thanks for the suggestions. I've made some of the changes. 

Regarding 2, 4, 6: I don't think we should add "failed" to attempt since I believe it adds confusion. "Attempts" already includes the idea that if people are failing. If they were able to login we wouldn't call it an attempt. And adding "failed" isn't correct since I believe it is including a successful login in the count.

**Comment by herbdool** [comment]
@opi @klonos please retest.

**Comment by quicksketch** [comment]
I filed a review with more `login` vs `log in` changes, but reading over my suggestions, I think things are actually fine as-is. https://github.com/backdrop/backdrop/pull/2187 looks RTBC to me. I'll leave this for others to review before merging.

**Comment by herbdool** [comment]
@quicksketch thanks for the review. I had tried to be consistent so that as a verb it's two words "log in" and one word "login" as an adjective or noun. Sounds like you are now ok with it.

**Comment by klonos** [comment]
OK, just tested again. This is looking good (I did not see any need to change `log in` <=> `login`), but still a few nits:

1. I agree with @quicksketch's request to change the description text that is shown in `/admin/config/people` to be a full sentence (including period at the end).

2. The "Users may log in using" and "Identify users attempting to log in, using" labels could use semicolons at the end.

3. Either use "...BY IP AN ADDRESS" / "...BY A USER", or use  "...BY IP ADDRESS" / "...BY USER" in the fieldset labels.

4. Seperate the single sentence referring to the per-user limit, and the purpose of the per-IP setting, add a few commas here and there, and generally make the help text less ambiguous:

    > Do not allow any login from the current user's IP, if the limit has been reached. This is independent of the per-user limit (see setting below). It is used to prevent malicious login attempts from one IP using many different user accounts. By default the limit is set to a reasonably high value, to account for legitimate use cases, such as multiple users from the same institution logging in from a common IP.

4. I still think that the second help/description should be simplified to match the first:

    > Do not allow further login attempts for the same user account, if the limit has been reached.

    ...if not, then because we allow mins/hours/day, "every X hours." should perhaps be "per specified amount of time.".

5. "more/less likely to lock out users." -> "to lock users out."

I'd hate to be holding up this PR because of these UI text changes, so feel free to merge, and I will file a follow-up issue/PR for these.

**Comment by herbdool** [comment]
1. Already did that.
2. That's not done anywhere else. See e.g. `admin/config/people/settings`.
3. Done
4. Better if you file a follow-up.
5. Changed to "per specified time period".
6. I prefer "lock out users". Either is fine in English.

**Comment by klonos** [comment]
1. Hmm, this is what I see in the PR sandbox:

  <img width="554" alt="screen shot 2018-07-28 at 9 38 02 pm" src="https://user-images.githubusercontent.com/2423362/43356152-ae1cf780-92ae-11e8-9300-fe69aa5bc3ef.png">

2. Hmm, you are absolutely right.
3. 👍 
4 & 5. Will file a follow-up.
6. English is not my mother tongue, so I go by what I recall having been taught, or by what I have heard/read. Folding ♠️ 😄

**Comment by klonos** [comment]
...never mind 1. ...cache clear fixed things 😄

**Comment by quicksketch** [comment]
Looks great. Thanks @herbdool for your work  on this! Thanks @klonos and @opi for the reviewing! Merged https://github.com/backdrop/backdrop/pull/2187 into 1.x for 1.11.0. Woot!
